### ISSUE TITLE
Can't start any two docker-compose environments after update to Windows 20H2

### ISSUE THREAD

**Issue by stephen-turner** [microsoft/Windows-Containers#140]
This was originally filed by user @marcusschroeder as docker/for-win#9999, but as it can be reliably reproduced on Windows 2010 (maybe 2004?) but not 1903, and as it only happens with Windows containers, we believe that it is a bug in the windows container code.

Original report:

> Expected behavior
> On my computer I use a couple of docker-compose environments with windows containers in parallel. Before the windows update to 2004 or 20H2 it was no problem to start several environments with
> docker-compose up either manually or programmatically. It didn't matter if it was the same env with a different name or a completely different one.
> 
> Actual behavior
> Now, when starting any two docker compose environments, the second one gets stuck in start up until the first is stopped.
> 
> Information
> Is it reproducible? Yes, even on another computer.
> Is the problem new? Yes, it appeared after windows update 1903 to 20H2
> Did the problem appear with an update? Yes
> Windows Version: Windows 10 Pro 20H2
> Docker Desktop Version: 2.5.0.1 / 2.4.0.0 / 2.3.0.4 / 2.5.1.0 (experimental) / 3.0.0.0
> 
> The problem appeared after a Windows update from 1903 to 2004 or 20H2 respectively.
> 
> I have tried:
> 
> different Docker Desktop versions (see above) to no avail.
> WSL2 and LCOW
>
> Steps to reproduce the behavior
> Use the following docker-compose.yml:
> 
```
version: "2.4"


services:
  service_a:
    image: mcr.microsoft.com/windows/servercore/iis
    ports:
      - 80


  service_b:
    image: mcr.microsoft.com/windows/servercore/iis
    ports:
      - 80
```
> 1. Use Docker Desktop with Windows Container Backend
> 2. Start first environment with above docker-compose.yml: docker-compose -p first up -d
> 3. Wait until environment is up (5-10 seconds)
> 4. Start second environment: docker-compose -p second up -d
> At this point docker-compose appears to be stuck and one of the services show status "created" in docker ps -a
> 5. Stop first environment in separate terminal docker-compose -p first down
> 6. Now, the command from 5. continues and the environment becomes healthy and the service status is "running"
> Furthermore:
> 
> if I remove the "ports" from the docker-compose.yml this issue does not occur.
> rolling back to windows 1903 also fixed the issue for now until the windows update is applied again

My colleague @StefanScherer has reproduced it without docker compose, but with a second nat network as follows:
```
$ docker network create -d nat first
10851710aef0c9645393c78f5480cc9d8c2309b079e4d8d82f70c9c6f1ee064f
$ docker run -d --network first -p 8004:80 mcr.microsoft.com/windows/servercore/iis
e89d12002a50b22b3628ddaca1ac06e4b34728ea803c7d5250d84721ddf993bf
$ docker run -d --network first -p 8005:80 mcr.microsoft.com/windows/servercore/iis
93adddb1f6213d21d0b8138b3fd47784938c238fa75256c31f7291c8713c57aa

$ docker network create -d nat second
e30d9c556d386408137db45222dc47d989e4c0d49a7f3a051f56ee93fa18c912
$ docker run -d --network second -p 8006:80 mcr.microsoft.com/windows/servercore/iis
f05233268a2e826f79f323659bf1e3a15333cd5d440117e053ccf04ceaf1a2c8
$ docker run -d --network second -p 8007:80 mcr.microsoft.com/windows/servercore/iis
3ea1546125d05b655e89913b5dd6677e3aa58d2e2d6b4f0fe93b1ff091ebb08e
```
> The Docker Cli of the last container does not return to the shell prompt, and in Docker Dashboard the fourth container is in CREATED state. 
> When I kill one of the first containers (e89d12) then the Docker Cli shows this error message
```
docker: Error response from daemon: failed to create endpoint serene_bhabha on network second: failed during hnsCallRawResponse: hnsCall failed in Win32: The specified port already exists. (0x803b0013).
```
> The Port 8007 was not used before.

**Comment by vrapolinario** [comment]
I'll get this issue assigned to the proper team soon.

**Comment by ghost** [comment]
This issue has been open for 30 days with no updates.
, please provide an update or close this issue.

**Comment by ghost** [comment]
This issue has been open for 30 days with no updates.
, please provide an update or close this issue.

**Comment by fady-azmy-msft** [comment]
Sorry for the delay @stephen-turner, is this still an issue you're facing?

**Comment by stephen-turner** [comment]
Thanks for reaching out. I assume it's still present unless some effort has been made to fix it, but I'm not actually the original reporter, I just transferred it from https://github.com/docker/for-win/issues/9999. The OR is @marcusschroeder.

**Comment by vrapolinario** [comment]
I don't know what the status of this is since I moved teams. Changed the assignment to @fady-azmy-msft.

**Comment by fady-azmy-msft** [comment]
Apologies for the radio silence. We have been able to reproduce the issue, and are looking into this. I've also created an internal ticket (41698313) to track this internally.

**Comment by logivity** [comment]
Is there going to be update on this?

**Comment by fady-azmy-msft** [comment]
@MikeZappa87 any updates on this?

**Comment by ntrappe-msft** [comment]
The team working on this has indicated that they're aware of the problem but have had to prioritize other tasks first. We'll update when we can.

**Comment by ntrappe-msft** [comment]
Good news coming soon 🎉

**Comment by jag-eagle-technology** [comment]
@ntrappe-msft super excited to hear that good news - any updates? 
We're keen to get this resolved as it's a blocker for our current CI/CD process & for the clients using our compose orchestrated software.

**Comment by ntrappe-msft** [comment]
Hi, I totally understand how important this is to fix asap. @MikeZappa87 do you have any updates? You know more about the status of this Issue than I do.

**Comment by ezw21** [comment]
I'm also very keen to get this fixed. Has there been any progress - this is pretty core functionality.

**Comment by ntrappe-msft** [comment]
Reassigning Issue to another dev.

**Comment by sikhness** [comment]
Has there been any progress on this?

**Comment by sikhness** [comment]
> Good news coming soon 🎉

Has there been any progress on this? It's been over half a year since "good news" was supposedly coming soon.

**Comment by sikhness** [comment]
> Good news coming soon 🎉

@ntrappe-msft, has there been any progress on this? It's been over half a year since "good news" was supposedly coming soon.

**Comment by adrianm-msft** [comment]
@sikhness, we're actively working on this. It’s taking a bit longer than expected, will post any updates here.

**Comment by sikhness** [comment]
Thanks @adrianm-msft, we're eagerly awaiting news!
Do you have any workarounds in the meantime as this feature is pretty fundamental.

**Comment by adrianm-msft** [comment]
@sikhness, it looks like rolling back to windows 1903 temporarily fixes the issue until a windows update is applied again.

**Comment by sikhness** [comment]
Thanks @adrianm-msft, this won't be possible in my environment as my host is running Windows Server 2022 and I don't think it can rollback to a version that old.
Eagerly awaiting the fix for this.

**Comment by Jens-G** [comment]
Using 2019 also helps. Not that I would recommend that, but indeed it still works there.
Being hit by it myself after moving my environment to Server 2022, so a fix would really be appreciated.

**Comment by marceliwac** [comment]
I would love to see this issue resolved as well. I've spent countless hours trying to debug the application logic and the docker setup on my Windows Server 2022 machine only to find that the problem is in fact the above.

For those looking for a temporary workaround (and I appreciate that this won't apply to more complex setups!), simplifying the networking stack in the compose file resolved all my issues with the containers sporadically hanging during startup, having to restart the docker service and in some cases the entire system etc. Specifically, I removed the custom networks and customised the default one instead to prevent having multiple networks (default network was being created automatically). I'm only mentioning this here to aid those who also struggled with the same problem in a hope it gets more visibility online.

**Comment by sikhness** [comment]
> I would love to see this issue resolved as well. I've spent countless hours trying to debug the application logic and the docker setup on my Windows Server 2022 machine only to find that the problem is in fact the above.
> 
> For those looking for a temporary workaround (and I appreciate that this won't apply to more complex setups!), simplifying the networking stack in the compose file resolved all my issues with the containers sporadically hanging during startup, having to restart the docker service and in some cases the entire system etc. Specifically, I removed the custom networks and customised the default one instead to prevent having multiple networks (default network was being created automatically). I'm only mentioning this here to aid those who also struggled with the same problem in a hope it gets more visibility online.

Hey @marceliwac,
Would you be able to provide an example of the workaround that you created?  What I'm currently doing is that in my compose file, I'm using the nat network as an external network so that it attaches it to the default nat network created by Windows Containers.  The problem with that though is that when you restart your machine, internally it seems Windows decides to recreate the default nat network each time so it's internal ID changes.  Because of that, any compose files (and thus their already deployed containers) pointing to the older ID prior to the restart, fail to start.  So I currently had to create and run a script on each startup that basically brings down all of my compose configs, then recreates them (to get the new network IDs on each boot).

**Comment by marceliwac** [comment]
@sikhness I'm afraid the workaround I use wouldn't be of help in this case. All I do is change the `network` block of docker-compose.yml to customise the `default` network, rather than create a new one.

If I recall correctly, when looking for the solution to this issue I stumbled upon a few threads which mentioned the same (or adjacent) issues you are facing. The solution involved setting a static MAC address for the HyperV NAT Virtual Adapter. I'm sorry I cannot give you anything more concrete, but here are few links that might be worth looking at. They are not exactly answers to your question but might give you some ideas:

https://superuser.com/questions/1701567/how-to-add-a-static-ip-adress-to-a-virtual-machine-in-hyper-v-to-stop-changing-t
https://superuser.com/questions/1815670/change-mac-adress-virtual-switch-windows-server-with-hyper-v

**Comment by Jens-G** [comment]
> please provide an update

That woulkd be indeed awesome. It seems as if this breaks a lot of stuff for enough people.

**Comment by adrianm-msft** [comment]
Hi, we'll be able to provide updates in a few days.

**Comment by sikhness** [comment]
Hi @adrianm-msft, we're eagerly awaiting good news! Also hoping whatever changes are being made to fix this issue would be present in Windows Server 2022 and the upcoming Windows Server 2025.

**Comment by adrianm-msft** [comment]
A fix for the issue is planned for the November release of Windows Server 2025. However, the fix for Windows Server 2022 will take a bit longer.

**Comment by Jens-G** [comment]
Sorry to ruin your nerves but how long is longer?

**Comment by adrianm-msft** [comment]
The fix for Windows Server 2022 is scheduled for the February release.

**Comment by adrianm-msft** [comment]
We've implemented a fix for the issue on Windows Server 2025. Has anyone had a chance to validate the fix? Any feedback would be greatly appreciated!

**Comment by leojth** [comment]
Hi @adrianm-msft, unfortunately it doesn't work on Windows Server 2025 either. Exact same behavior: starting third container will hang at command prompt and Host Network Service goes crazy with CPU. This is very easy to reproduce using the steps in https://github.com/microsoft/Windows-Containers/issues/486 (closed as a duplicate).

I did a fresh install of Windows Server 2025, with all the latest updates (OS Build 26100.2605), enabled Containers and used the latest Docker Engine (27.4.0).

Actually it is a bit worse in Windows Server 2025 since I can't recover by stopping any of the other containers. That worked in 2022. The only way to recover now is to reboot the server.

**Comment by adrianm-msft** [comment]
@leojth thanks for the update! We're looking into it now.

**Comment by Eisaichen** [comment]
@adrianm-msft Thanks for your work!
Unfortunately, I got the same result as leojth got.
I found this bug will not trigger if the container does not use a port. As the screenshot below shows, when no port is assgined, I can run 4 containers simultaneously, each with its own individual NAT network.
<details><summary>screenshot1</summary>
<img width="867" alt="Image" src="https://github.com/user-attachments/assets/d091e8e6-7d30-4aad-8a5c-abb5c7f78b74" />
</details>

But if I modify the configuration to assign a port for each container, then this will 100% trigger the bug.
<details><summary>screenshot2</summary>
<img width="867" alt="Image" src="https://github.com/user-attachments/assets/d3e4ca45-13a0-4519-acb0-79d1e81f8159" />
</details>
(Stuck when starting the 3rd container)

I tested on Server 2025 `26100.2605` with `process` isolation.
<details><summary>docker system info</summary>

```
Client:
 Version:    27.4.1
 Context:    default
 Debug Mode: false
 Plugins:
  compose: Docker Compose (Docker Inc.)
    Version:  v2.32.1
    Path:     C:\Users\Administrator\.docker\cli-plugins\docker-compose.exe

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 1
 Server Version: 27.4.1
 Storage Driver: windowsfilter
  Windows:
 Logging Driver: json-file
 Plugins:
  Volume: local
  Network: ics internal l2bridge l2tunnel nat null overlay private transparent
  Log: awslogs etwlogs fluentd gcplogs gelf json-file local splunk syslog
 Swarm: inactive
 Default Isolation: process
 Kernel Version: 10.0 26100 (26100.1.amd64fre.ge_release.240331-1435)
 Operating System: Microsoft Windows Server Version 24H2 (OS Build 26100.2605)
 OSType: windows
 Architecture: x86_64
 CPUs: 32
 Total Memory: 31.21GiB
 Name: WIN-AM29EO7G1RC
 ID: 332dbb75-cfdc-42a0-9a08-5b2fe1d470ed
 Docker Root Dir: C:\ProgramData\docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine
```
</details>

BTW, I assume Win 11 will get the fix at the same time as Server 2025? Since they share the same build number.
(Can't believe this issue is more than 3 years old and Microsoft still has not abandon the project)

**Comment by adrianm-msft** [comment]
We've done some testing on our end and seem to be facing issues as well - will keep you posted!

**Comment by adrianm-msft** [comment]
We've worked on resolving this issue, and it should now be fixed in the latest January release for WS2025. Has anyone had a chance to validate it? Thanks for your feedback!

**Comment by Eisaichen** [comment]
@adrianm-msft tested on server2025 `26100.2894`, it does not seem to fix the issue, behalf as same as previous.
<details><summary>screenshot</summary>
<img width="874" alt="Image" src="https://github.com/user-attachments/assets/563f1414-e560-4d73-b2e0-dfef603efeee" />
</details>

**Comment by soonoh** [comment]
I heard the latest release was 25.01D which is KB5050094. 

https://support.microsoft.com/en-us/topic/january-28-2025-kb5050094-os-build-26100-3037-preview-78fda0ea-79e9-468d-8a77-de7914ca1aef

You can find whether the 25.01D release installed on the machine using the PowerShell command "Get-Hotfix." After installing it, the OS build will be OS Build 26100.3037. The above screen shot seems installing only 25.01B.

**Comment by Eisaichen** [comment]
@soonoh The update you gave is currently still under preview. And it is currently only available on win11, not server 2025. I usually prefer not to install a preview update since Windows already has more than enough bugs. Also, according to adrianm, the fix is implemented on server2025.

<details>
<img width="204" alt="Image" src="https://github.com/user-attachments/assets/1d3199a8-7b59-40fa-a2d3-73888b809587" />

![Image](https://github.com/user-attachments/assets/ea4dc3ae-b0bf-4a89-9824-95dbec207879)
</details>

**Comment by soonoh** [comment]
I don't know why it is preview. I installed it on my WS2025 VM and there was no problem to install. Unfortunately,

<img width="638" alt="Image" src="https://github.com/user-attachments/assets/748cb080-c782-4992-bb0d-00f65644bfb5" />

**Comment by Eisaichen** [comment]
@soonoh You can literally see the "preview" in the URL you gave. Anyway since you have it, maybe you could go verify the fix and give us the result? Thanks in advance.

**Comment by Eisaichen** [comment]
Good news! This issue seems fixed on 26100.3194 which was pushed on yesterday.

**Comment by Eisaichen** [comment]
26100.3037 seems never reached Windows Server on Windows Update. I tested on 26100.3194 which was only pushed on yesterday.

<img width="200" alt="Image" src="https://github.com/user-attachments/assets/0e63a2ef-7692-49c1-932c-2b3130511a1f" />

Unfortunately, the result is the same as before. @adrianm-msft Does the patch worked when you test it internally?
<details>

<img width="665" alt="Image" src="https://github.com/user-attachments/assets/7d2de456-4279-41ea-8def-e766cf7118ac" />

</details>

**Comment by leojth** [comment]
Hi @adrianm-msft! I tested now on Windows Server 2025 build 26100.3476 that was released today, but unfortunately the problem still exists. Third container hangs and Host Network Service takes allot of CPU. No way to recover but to reboot the server.

**Comment by ezw21** [comment]
> The fix for Windows Server 2022 is scheduled for the February release.

Hi Adrian, thanks for looking into this. Is this issue resolved in WindowsServer 2022?

**Comment by leojth** [comment]
Hi @adrianm-msft, now tested with Windows Server 2025 build 26100.3775 but problem still exists. Do you have any update on when a fix can be expected?

We have basically stayed on Windows Server 2019 because of this issue but we really want to upgrade soon.

**Comment by adrianm-msft** [comment]
Hi @leojth, due to some unrelated issues in the TCPIP bugfix bundle, many fixes, including this one, have been delayed. We anticipate that it will take at least 1 to 2 months before this change is shipped.

**Comment by Jens-G** [comment]
> due to some unrelated issues in the TCPIP bugfix bundle, many fixes, including this one, have been delayed. We anticipate that it will take at least 1 to 2 months before this change is shipped.

Time is relative, isn't it?

**Comment by Jens-G** [comment]
Can someone please officially confirm that all the above is essentially BS and the issue is not intended at all to be fixed at any time? Then we can close this as WONTFIX and move on to Linux.

**Comment by leojth** [comment]
Soon 6 months since last status update and over 4 years since issue first raised. This is a **serious bug** that effectively stops many usages of native containers on Windows Server. It worked before, but not now. I think it is time to give this issue higher priority.

**Comment by mariotomek** [comment]
Hello, is there any progress on fixing this bug? Many of us have been waiting for this fix for a really long time. Would it be possible to raise the priority of this bug?
